compiler/src/iree/compiler/Codegen/Dialect/GPU/IR/IREEGPUOps.cpp

```diff
@@ -237,15 +235,67 @@ LogicalResult CoalescedGatherDMAOp::verify() {
     return emitOpError("input type must either be a tensor or a memref");
   }
 
-  if (hasTensor) {
-    if (!isa<RankedTensorType>(getIndices().getType()) ||
-        !isa<RankedTensorType>(getSource().getType())) {
-      return emitOpError("all operands must be tensors when init is a tensor");
+  auto initShape = cast<ShapedType>(initType).getShape();
+  auto sourceType = cast<ShapedType>(getSource().getType());
+
+  if (Value indices = getIndices()) {
+    // Gather mode:
+    if (!isa<VectorType>(indices.getType())) {
+      return emitOpError("indices must be a vector of index type");
+    }
+
+    auto indicesVectorType = cast<VectorType>(indices.getType());
+    if (!indicesVectorType.getElementType().isIndex()) {
+      return emitOpError("indices must be a vector of index type");
+    }
+
+    if (hasTensor) {
+      if (!isa<RankedTensorType>(getSource().getType())) {
+        return emitOpError("source must be a tensor when init is a tensor");
+      }
+    } else if (hasMemRef) {
+      if (!isa<MemRefType>(getSource().getType())) {
+        return emitOpError("source must be a memref when init is a memref");
+      }
+    }
+
+    if (sourceType.getRank() != 1) {
+      return emitOpError(
+          "source must be a 1-D tensor/memref when indices is present");
     }
-  } else if (hasMemRef) {
-    if (!isa<MemRefType>(getIndices().getType()) ||
-        !isa<MemRefType>(getSource().getType())) {
-      return emitOpError("all operands must be memrefs when init is a memref");
+
+    auto indicesShape = indicesVectorType.getShape();
+    if (indicesShape.size() > initShape.size()) {
+      return emitOpError(
+          "indices shape must be a prefix of or match the result/init shape");
+    }
+
+    for (size_t i = 0; i < indicesShape.size(); ++i) {
+      if (!ShapedType::isDynamic(indicesShape[i]) &&
+          !ShapedType::isDynamic(initShape[i]) &&
+          indicesShape[i] != initShape[i]) {
+        return emitOpError("indices shape must be a prefix of the result/init "
+                           "shape, mismatch at dim ")
+               << i;
+      }
+    }
+  } else {
+    // Copy mode:
+    if (hasTensor) {
+      if (!isa<RankedTensorType>(getSource().getType())) {
+        return emitOpError("source must be a tensor when init is a tensor");
+      }
+    } else if (hasMemRef) {
+      if (!isa<MemRefType>(getSource().getType())) {
+        return emitOpError("source must be a memref when init is a memref");
+      }
+    }
+
+    // In copy mode, source and init must have the same rank
```

Comment:
```suggestion
    // In copy mode, source and init must have the same rank.
```
See https://llvm.org/docs/CodingStandards.html#commenting